keel-sql/src/main/kotlin/com/netflix/spinnaker/keel/sql/SqlArtifactRepository.kt

```diff
@@ -433,6 +463,107 @@ class SqlArtifactRepository(
     }
   }
 
+  override fun pinEnvironment(deliveryConfig: DeliveryConfig, environmentArtifactPin: EnvironmentArtifactPin) {
+    with(environmentArtifactPin) {
+      val environment = deliveryConfig.environmentNamed(targetEnvironment)
+      val artifact = get(deliveryConfigName, reference, ArtifactType.valueOf(type.toUpperCase()))
+
+      sqlRetry.withRetry(WRITE) {
+        jooq.insertInto(ENVIRONMENT_ARTIFACT_PIN)
+          .set(ENVIRONMENT_ARTIFACT_PIN.ENVIRONMENT_UID, deliveryConfig.getUidFor(environment))
+          .set(ENVIRONMENT_ARTIFACT_PIN.ARTIFACT_UID, artifact.uid)
+          .set(ENVIRONMENT_ARTIFACT_PIN.ARTIFACT_VERSION, version)
+          .set(ENVIRONMENT_ARTIFACT_PIN.PINNED_AT, clock.millis())
+          .set(ENVIRONMENT_ARTIFACT_PIN.PINNED_BY, pinnedBy ?: "anonymous")
+          .set(ENVIRONMENT_ARTIFACT_PIN.COMMENT, comment)
```

Comment:
noob sql question: do we need to handle duplicate keys here, and update the time/pinned by / comment?

Reply:
Sure do! Great catch.